src/lib/shim/src/lib.rs

```diff
@@ -24,9 +34,191 @@ pub fn manager_shmem() -> Option<&'static ManagerShmem> {
+// stack guard. We assume 4k pages here but detect at runtime if this is too small.
+const SHIM_SIGNAL_STACK_GUARD_OVERHEAD: usize = 4096 * 2;
+
+mod tls_thread_signal_stack {
+    use super::*;
+
+    static THREAD_SIGNAL_STACK: ShimTlsVar<Cell<*mut core::ffi::c_void>> =
+        ShimTlsVar::new(|| Cell::new(core::ptr::null_mut()));
+
+    // Shouldn't need to make this very large, but needs to be big enough to run the
+    // managed process's signal handlers as well - possibly recursively.
+    //
+    // Stack space that's *never* used shouldn't ever become resident, but an
+    // occasional deep stack could force the pages to be resident ever after.  To
+    // mitigate that, we could consider `madvise(MADV_DONTNEED)` after running
+    // signal handlers, to let the OS reclaim the (now-popped) signal handler stack
+    // frames.
+    const SHIM_SIGNAL_STACK_MIN_USABLE_SIZE: usize = 1024 * 100;
+    const SHIM_SIGNAL_STACK_SIZE: usize =
+        SHIM_SIGNAL_STACK_GUARD_OVERHEAD + SHIM_SIGNAL_STACK_MIN_USABLE_SIZE;
+
+    /// # Safety
+    ///
+    /// See `sigaltstack(2)`
+    pub unsafe fn init() {
+        // Allocate
+        let new_stack = unsafe {
+            rustix::mm::mmap_anonymous(
+                core::ptr::null_mut(),
+                SHIM_SIGNAL_STACK_SIZE,
+                rustix::mm::ProtFlags::READ | rustix::mm::ProtFlags::WRITE,
+                rustix::mm::MapFlags::PRIVATE,
+            )
+        }
+        .unwrap();
+
+        // Save to thread-local, so that we can deallocate on thread exit
+        assert!(
+            THREAD_SIGNAL_STACK.get().replace(new_stack).is_null(),
+            "Allocated signal stack twice for current thread"
+        );
+
+        // Set up guard page
+        unsafe { rustix::mm::mprotect(new_stack, 4096, rustix::mm::MprotectFlags::empty()) }
+            .unwrap();
+
+        // Install via `sigaltstack`. The kernel will switch to this stack when
+        // invoking one of our signal handlers.
+        let stack_descriptor = linux_api::signal::stack_t {
+            ss_sp: new_stack,
+            ss_size: SHIM_SIGNAL_STACK_SIZE.try_into().unwrap(),
+            // Clear the alternate stack settings on entry to signal handler, and
+            // restore it on exit.  Otherwise a signal handler invoked while another
+            // is running on the same thread would clobber the first handler's stack.
+            // Instead we want the second handler to push a new frame on the alt
+            // stack that's already installed.
+            ss_flags: linux_api::signal::SigAltStackFlags::SS_AUTODISARM.bits(),
+        };
+        unsafe {
+            linux_api::signal::sigaltstack(Some(&stack_descriptor), None).unwrap();
+        }
+    }
+
+    /// # Safety
+    ///
+    /// The current thread must exit before:
+    /// * Another thread runs.
+    /// * That thread also frees its stack.
+    /// * This thread runs again on the signal stack (e.g. by handling a new signal).
```

Comment:
These safety requirements aren't clear to me. Maybe it should say something like "After calling `free()`, the current thread must exit before"? Then what does it mean by "the current thread must exit before this thread runs again on the signal stack" (how can the thread run if it's already exited)?